byron/semantics/executable-spec/src/Control/State/Transition/Generator.hs

```diff
@@ -50,88 +62,91 @@ import Control.State.Transition.Trace
+      Left _pf  -> trace n
       -- Applying an initial rule with an environment and state will simply
       -- validate that state, so we do not care which state 'applySTS' returns.
-      Right st -> genTrace env st (sigGen @s)
+      Right st -> genTrace n env st (sigGen @s)
 
 -- | Return a (valid) trace generator given an initial state, environment, and
 -- signal generator.
+--
 genTrace
   :: forall s
    . STS s
-  => Environment s
+  => Int
+  -- ^ Trace upper bound. This will be linearly scaled as a function of the
+  -- generator size.
+  -> Environment s
+  -- ^ Environment, which remains constant in the system.
   -> State s
+  -- ^ Initial state.
   -> (Environment s -> State s -> Gen (Signal s))
+  -- ^ Signal generator. This generator relies on an environment and a state to
+  -- generate a signal.
   -> Gen (Trace s)
-genTrace env st aSigGen = Gen.sized $ \d -> mkTrace env st <$> go d st []
-  where
-    go d sti acc =
-      Gen.frequency [ (5, return acc)
-                    -- The probability of continue with the recursion depends
-                    -- on the size parameter of the generator. Here the
-                    -- constant factor is determined ad-hoc.
-                    , (unSize d * 2, do
-                          mStSig <- genSigSt @s env sti aSigGen
-                          case mStSig of
-                            Nothing ->
-                              go d sti acc
-                            Just (stNext, sig) ->
-                              go d stNext ((stNext, sig): acc)
-                      )
-                    ]
-  -- An alternate way to generate a trace of the size of the generator might
-  -- be:
-  --
-  -- >>>  go 0 _   acc = return acc
-  -- >>>  go d sti acc = do
-  -- >>>    mStSig <- genSigSt @s env sti aSigGen
-  -- >>>    case mStSig of
-  -- >>>      Nothing ->
-  -- >>>        go (d - 1) sti acc
-  -- >>>      Just (stNext, sig) ->
-  -- >>>        go (d - 1) stNext ((stNext, sig): acc)
+genTrace ub env st0 aSigGen = do
+  -- Generate the initial size of the trace, but don't shrink it (notice the
+  -- use of 'integral_') since we will be shrinking the traces manually (so it
+  -- doesn't make sense to shrink the trace size).
   --
+  -- Note that the length of the resulting trace might be less than the
+  -- generated value if invalid signals (according to some current state) are
+  -- generated in 'loop'.
+  n <- integral_ $ Range.linear 0 ub
+  withGenT (mapGenT (TreeT . interleaveSigs . runTreeT)) $ loop n st0 []
+  where
+    loop
+      :: Int
+      -> State s
+      -> [TreeT (MaybeT Identity) (Signal s)]
+      -> Gen [TreeT (MaybeT Identity) (Signal s)]
+    loop 0 _ acc = pure acc
+    loop d sti acc = do
+      sigTree :: TreeT (MaybeT Identity) (Signal s)
+        <- toTreeMaybeT $ aSigGen env sti
+      let
+        --  Take the root of the next-state signal tree.
+        mSig = fmap nodeValue $ runIdentity $ runMaybeT (runTreeT sigTree)
```

Comment:
You can simplify this with [runDiscardEffect](http://hackage.haskell.org/package/hedgehog-1.0/docs/Hedgehog-Internal-Gen.html#v:runDiscardEffect)

Reply:
Indeed! I replaced this with 

```haskell
mSig = treeValue <$> runDiscardEffect sigTree
```

:pray: